go/vt/mysqlctl/replication.go

```diff
@@ -254,14 +272,50 @@ func (mysqld *Mysqld) SetReadOnly(on bool) error {
 }
 
 // SetSuperReadOnly set/unset the super_read_only flag
-func (mysqld *Mysqld) SetSuperReadOnly(on bool) error {
+func (mysqld *Mysqld) SetSuperReadOnly(on bool) (ResetSuperReadOnlyFunc, error) {
+	//  return function for switching `OFF` super_read_only
+	var returnFunc ResetSuperReadOnlyFunc
+	var resetFunc = func() error {
+		query := "SET GLOBAL super_read_only = 'OFF'"
+		err := mysqld.ExecuteSuperQuery(context.TODO(), query)
+		return err
+	}
+
+	//  return function for switching `ON` super_read_only
+	var setFunc = func() error {
+		query := "SET GLOBAL super_read_only = 'ON'"
+		err := mysqld.ExecuteSuperQuery(context.TODO(), query)
+		return err
+	}
+
+	//var err error = nil
+	superReadOnlyEnabled, err := mysqld.IsSuperReadOnly()
+	if err != nil {
+		return nil, err
+	}
+
+	// if non-idempotent then set the right call-back
+	if on && !superReadOnlyEnabled {
+		returnFunc = resetFunc
+	}
+	if !on && superReadOnlyEnabled {
+		returnFunc = setFunc
+	}
+
 	query := "SET GLOBAL super_read_only = "
 	if on {
-		query += "ON"
+		query += "'ON'"
 	} else {
-		query += "OFF"
+		query += "'OFF'"
 	}
-	return mysqld.ExecuteSuperQuery(context.TODO(), query)
+	if err := mysqld.ExecuteSuperQuery(context.TODO(), query); err != nil {
+		return nil, err
+	}
+	if superReadOnlyEnabled {
```

Comment:
We only return `returnFunc` if `superReadOnlyEnabled` is true. So when is `resetFunc()` ever returned as the `returnFunc`? We should add some unit tests around this function to clarify and validate the logic.

```
	if on && !superReadOnlyEnabled {
		returnFunc = resetFunc
	}
	if !on && superReadOnlyEnabled {
		returnFunc = setFunc
	}
```

Reply:
we return 'reset' or 'set' function as returnFunc with following logic
        
        // we are asked to turn on super_read_only but original value is false
        // then return resetFunc, that can be used as defer by caller
	if on && !superReadOnlyEnabled {
		returnFunc = resetFunc
	}
        // we are asked to turn off super_read_only but original value is true
        // then return setFunc, that can be used as defer by caller
	if !on && superReadOnlyEnabled {
		returnFunc = setFunc
	}

Reply:
for any other case return function is going to return nil .. since there is no internal state change.

Reply:
fixed. yes there was bug :)